File: app/src/main/java/ohi/andre/consolelauncher/commands/main/raw/orientation.java
package ohi.andre.consolelauncher.commands.main.raw;

import android.app.Activity;
import android.content.pm.ActivityInfo;

import java.util.Locale;

import ohi.andre.consolelauncher.LauncherActivity;
import ohi.andre.consolelauncher.R;
import ohi.andre.consolelauncher.commands.CommandAbstraction;
import ohi.andre.consolelauncher.commands.ExecutePack;
import ohi.andre.consolelauncher.managers.settings.LauncherSettings;
import ohi.andre.consolelauncher.managers.xml.options.Behavior;

public class orientation implements CommandAbstraction {

    static final String MODE_LANDSCAPE = "landscape";
    static final String MODE_PORTRAIT = "portrait";
    static final String MODE_AUTO = "auto";

    private static final String VALUE_LANDSCAPE = "0";
    private static final String VALUE_PORTRAIT = "1";
    private static final String VALUE_AUTO = "2";

    @Override
    public String exec(ExecutePack pack) {
        String input = pack.getString();
        return apply(pack, input);
    }

    static String apply(ExecutePack pack, String input) {
        String mode = input == null ? "" : input.trim().toLowerCase(Locale.US);
        if (mode.length() == 0 || "status".equals(mode) || "-status".equals(mode)) {
            return currentOrientation() + "\nUsage: orientation portrait|landscape|auto";
        }

        if (MODE_LANDSCAPE.equals(mode) || VALUE_LANDSCAPE.equals(mode)) {
            setOrientation(pack, VALUE_LANDSCAPE, ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
            return "Landscape mode enabled. Use orientation portrait or orientation auto to change it.";
        }

        if (MODE_PORTRAIT.equals(mode) || VALUE_PORTRAIT.equals(mode)) {
            setOrientation(pack, VALUE_PORTRAIT, ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
            return "Portrait mode enabled. Use orientation landscape or orientation auto to change it.";
        }

        if (MODE_AUTO.equals(mode)
                || VALUE_AUTO.equals(mode)
                || "autorotate".equals(mode)
                || "auto-rotate".equals(mode)) {
            setOrientation(pack, VALUE_AUTO, ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
            return "Auto-rotate enabled. Use orientation portrait or orientation landscape to pin it.";
        }

        return "Unknown orientation: " + input + "\nUsage: orientation portrait|landscape|auto";
    }

    private static void setOrientation(ExecutePack pack, String value, int fallbackOrientation) {
        LauncherSettings.set(pack.context, Behavior.orientation, value);

        if (pack.context instanceof LauncherActivity) {
            ((LauncherActivity) pack.context).applyOrientationPreference();
        } else if (pack.context instanceof Activity) {
            ((Activity) pack.context).setRequestedOrientation(fallbackOrientation);
        }
    }

    private static String currentOrientation() {
        int value = LauncherSettings.getInt(Behavior.orientation);
        if (value == 0) {
            return "Orientation: landscape";
        }
        if (value == 1) {
            return "Orientation: portrait";
        }
        if (value == 2) {
            return "Orientation: auto";
        }
        return "Orientation: " + value;
    }

    @Override
    public int[] argType() {
        return new int[] {CommandAbstraction.PLAIN_TEXT};
    }

    @Override
    public int priority() {
        return 4;
    }

    @Override
    public int helpRes() {
        return R.string.help_orientation;
    }

    @Override
    public String onArgNotFound(ExecutePack info, int index) {
        return info.context.getString(R.string.help_orientation);
    }

    @Override
    public String onNotArgEnough(ExecutePack info, int nArgs) {
        return currentOrientation() + "\nUsage: orientation portrait|landscape|auto";
    }
}

File: app/src/main/java/ohi/andre/consolelauncher/commands/main/raw/orientation.kt
package ohi.andre.consolelauncher.commands.main.raw

import java.util.Locale
import ohi.andre.consolelauncher.LauncherActivity
import ohi.andre.consolelauncher.R
import ohi.andre.consolelauncher.commands.CommandAbstraction
import ohi.andre.consolelauncher.commands.ExecutePack
import ohi.andre.consolelauncher.managers.settings.LauncherSettings
import ohi.andre.consolelauncher.managers.xml.options.Behavior

class orientation : CommandAbstraction {
    override fun exec(pack: ExecutePack): String {
        val input = pack.getString()
        return apply(pack, input)
    }

    override fun argType(): IntArray = intArrayOf(CommandAbstraction.PLAIN_TEXT)

    override fun priority(): Int = 4

    override fun helpRes(): Int = R.string.help_orientation

    override fun onArgNotFound(pack: ExecutePack, indexNotFound: Int): String =
        pack.context.getString(R.string.help_orientation)

    override fun onNotArgEnough(pack: ExecutePack, nArgs: Int): String =
        pack.context.getString(R.string.command_orientation_usage_orientation_portrait_landscape_auto_fbc79, currentOrientation(pack.context))

    companion object {
        const val MODE_LANDSCAPE = "landscape"
        const val MODE_PORTRAIT = "portrait"
        const val MODE_AUTO = "auto"

        private const val VALUE_LANDSCAPE = "0"
        private const val VALUE_PORTRAIT = "1"
        private const val VALUE_AUTO = "2"

        @JvmStatic
        fun apply(pack: ExecutePack, input: String?): String {
            val mode = input?.trim()?.lowercase(Locale.US) ?: ""
            if (mode.isEmpty() || mode == "status" || mode == "-status") {
                return pack.context.getString(R.string.command_orientation_usage_orientation_portrait_landscape_auto_fbc79, currentOrientation(pack.context))
            }

            if (mode == MODE_LANDSCAPE || mode == VALUE_LANDSCAPE) {
                setOrientation(pack, VALUE_LANDSCAPE)
                return pack.context.getString(R.string.command_orientation_landscape_preference_saved_re_t_ui_will_us_4c485)
            }

            if (mode == MODE_PORTRAIT || mode == VALUE_PORTRAIT) {
                setOrientation(pack, VALUE_PORTRAIT)
                return pack.context.getString(R.string.command_orientation_portrait_preference_saved_re_t_ui_will_kee_d0e0b)
            }

            if (mode == MODE_AUTO || mode == VALUE_AUTO || mode == "autorotate" || mode == "auto-rotate") {
                setOrientation(pack, VALUE_AUTO)
                return pack.context.getString(R.string.command_orientation_auto_orientation_preference_saved_re_t_ui_218c8)
            }

            return pack.context.getString(R.string.command_orientation_unknown_orientation_usage_orientation_port_4d4ad, input)
        }

        private fun setOrientation(pack: ExecutePack, value: String) {
            LauncherSettings.set(pack.context, Behavior.orientation, value)

            if (pack.context is LauncherActivity) {
                (pack.context as LauncherActivity).applyOrientationPreference()
            }
        }

        private fun currentOrientation(context: android.content.Context): String {
            val value = LauncherSettings.getInt(Behavior.orientation)
            if (value == 0) {
                return context.getString(R.string.command_detail_orientation_orientation_landscape_eb4bf)
            }
            if (value == 1) {
                return context.getString(R.string.command_detail_orientation_orientation_portrait_52aca)
            }
            return if (value == 2) context.getString(R.string.command_detail_orientation_orientation_auto_4c066) else context.getString(R.string.command_detail_orientation_orientation_919b3, value)
        }
    }
}
